### PR TITLE
Add certora rules and workflow verification step

### PR DESCRIPTION
**Pull request by toninorair**

### CODE REVIEW

File: certora/harness/MTokenHarness.sol [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    constructor(address ttgRegistrar_, address minterGateway_) MToken(ttgRegistrar_, minterGateway_) {}
+
+    /******************************************************************************************************************\
+    |                                                     Getters                                                      |
+    \******************************************************************************************************************/
+
+
+    function getLatestIndexInMToken() public view returns (uint128) {
+        return latestIndex;
+    }
```
Comment (deluca-mike) [inline certora/harness/MTokenHarness.sol:18]: Why is this needed, isn't `latestIndex` already a public getter? And why is there a redundant `InMToken` suffix in the function name?

File: certora/harness/MTokenHarness.sol [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    function getLatestRateInMToken() public view returns (uint32) {
+        return _latestRate;
+    }
```
Comment (deluca-mike) [inline certora/harness/MTokenHarness.sol:22]: Why is this needed, isn't `earnerRate()` already a public getter? And why is there a redundant `InMToken` suffix in the function name?

File: certora/harness/MTokenHarness.sol [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    function getLatestUpdateTimestampInMToken() public view returns (uint40) {
+        return latestUpdateTimestamp;
+    }
```
Comment (deluca-mike) [inline certora/harness/MTokenHarness.sol:26]: Why is this needed, isn't `latestUpdateTimestamp` already a public getter? And why is there a redundant `InMToken` suffix in the function name?

File: certora/harness/MTokenHarness.sol [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    function getIsEarning(address account_) public view returns (bool) {
+        return _balances[account_].isEarning;
+    }
```
Comment (deluca-mike) [inline certora/harness/MTokenHarness.sol:30]: Why is this needed, isn't `isEarning` already a public getter?

File: certora/harness/MTokenHarness.sol [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    function getInternalBalanceOf(address account_) public view returns (uint240) {
+        return _balances[account_].rawBalance;
+    }
+
+    function getEarnerRate() public view returns (uint32) {
```
Comment (deluca-mike) [inline certora/harness/MTokenHarness.sol:36]: Should be called `getRate` or `getInternalRate`

File: certora/harness/MinterGatewayHarness.sol [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        return uint240(IMToken(mToken).totalSupply()); 
+    }
+
+    /******************************************************************************************************************\
+    |                                                     Getters                                                      |
+    \******************************************************************************************************************/
+
+    function getLatestIndexInMinterGateway() public view returns (uint128) {
+        return latestIndex;
+    }
```
Comment (deluca-mike) [inline certora/harness/MinterGatewayHarness.sol:21]: Why is this needed, isn't `latestIndex` already a public getter? And why is there a redundant `InMinterGateway` suffix in the function name?

File: certora/harness/MinterGatewayHarness.sol [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    function getLatestRateInMinterGateway() public view returns (uint32) {
+        return _latestRate;
+    }
```
Comment (deluca-mike) [inline certora/harness/MinterGatewayHarness.sol:25]: Why is this needed, isn't `minterRate` already a public getter? And why is there a redundant `InMinterGateway` suffix in the function name?

File: certora/harness/MinterGatewayHarness.sol [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+    function getLatestUpdateTimestampInMinterGateway() public view returns (uint40) {
+        return latestUpdateTimestamp;
+    }
```
Comment (deluca-mike) [inline certora/harness/MinterGatewayHarness.sol:29]: Why is this needed, isn't `latestUpdateTimestamp` already a public getter? And why is there a redundant `InMinterGateway` suffix in the function name?

File: certora/harness/MinterGatewayHarness.sol [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    function getPrincipalAmountRoundedDown(uint240 amount_) public view returns (uint112 principalAmount_) {
+        return _getPrincipalAmountRoundedDown(amount_);
+    }
+
+    function getPrincipalAmountRoundedUp(uint240 amount_) public view returns (uint112 principalAmount_) {
+        return _getPrincipalAmountRoundedUp(amount_);
+    }
+
+    function getMinterRate() public view returns (uint32) {
```
Comment (deluca-mike) [inline certora/harness/MinterGatewayHarness.sol:39]: Should be called `getRate` or `getInternalRate`.

File: certora/harness/MinterGatewayHarness.sol [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+import "../../src/MinterGateway.sol";
+
+contract MinterGatewayHarness is MinterGateway {
+    constructor(address ttgRegistrar_, address mToken_) MinterGateway(ttgRegistrar_, mToken_) {}
+
+    function totalMSupply() public view returns (uint240) {
+        // Note this is also used to determine the total M supply in
+        // functions implemented in MinterGatway
+        return uint240(IMToken(mToken).totalSupply()); 
+    }
```
Comment (deluca-mike) [inline certora/harness/MinterGatewayHarness.sol:13]: What is the purpose of this function? Why can't the user of this harness just call `totalSupply` at `mToken` directly?

**Review by deluca-mike — DISMISSED** [review]
Can't go over all the specs right now, but once merged, we will have time to run and review them ourselves.

File: .github/workflows/certora_verification.yml [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+      - main
+
+  workflow_dispatch:
+
+jobs:
+  verify:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
```
Comment (PierrickGT) [inline .github/workflows/certora_verification.yml:18]:
```suggestion
      - uses: actions/checkout@v4
```

File: .github/workflows/certora_verification.yml [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        with:
+          submodules: recursive
+
+      - name: Install python
+        uses: actions/setup-python@v2
```
Comment (PierrickGT) [inline .github/workflows/certora_verification.yml:23]:
```suggestion
        uses: actions/setup-python@v5
```

File: .github/workflows/certora_verification.yml [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        with: { python-version: 3.9 }
+
+      - name: Install java
+        uses: actions/setup-java@v1
```
Comment (PierrickGT) [inline .github/workflows/certora_verification.yml:27]:
```suggestion
        uses: actions/setup-java@v4
```

**Review by PierrickGT — APPROVED** [review]

**Review by deluca-mike — APPROVED** [review]
